File: scripts/update-readme.js
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import process from "process";
import fg from "fast-glob";
import { XMLParser } from "fast-xml-parser";
import { Octokit } from "@octokit/rest";

const MARKER_START = "<!-- ideas:start -->";
const MARKER_END = "<!-- ideas:end -->";
const README_PATH = "README.md";
const ATTR_CACHE_PATH = ".idea_attribution.json";

const FOCUS_MAP = {
    cure: "🧪 Cause and Cure",
    disaster: "🆘 Disaster and Community Support",
    education: "🎓 Education and Youth Services",
    sustainability: "🌱 Sustainability and Decarbonization",
    other: "🧩 Other",
}

// Matches any folder depth to the update path in scope
const GLOB_PATTERN = "**/update/x_snc_hack4good_0_hack4good_proposal_*.xml";

const repoFull = process.env.GITHUB_REPOSITORY || "";
const [OWNER, REPO] = repoFull.split("/");

// Octokit (uses the ephemeral GITHUB_TOKEN in Actions)
const octokit = new Octokit({
  auth: process.env.GITHUB_TOKEN || process.env.GH_TOKEN || undefined,
  userAgent: "hack4good-readme-bot"
});

// --- utilities

function friendlyFocus(value) {
  if (!value) return "—";
  const key = String(value).trim().toLowerCase();
  return FOCUS_MAP[key] ?? key.replace(/_/g, " ").replace(/\b\w/g, s => s.toUpperCase());
}

function loadCache() {
  try {
    if (fs.existsSync(ATTR_CACHE_PATH)) {
      return JSON.parse(fs.readFileSync(ATTR_CACHE_PATH, "utf-8"));
    }
  } catch { /* ignore */ }
  return {};
}

function saveCache(cache) {
  fs.writeFileSync(ATTR_CACHE_PATH, JSON.stringify(cache, null, 2) + "\n", "utf-8");
}

function readReadme() {
  if (fs.existsSync(README_PATH)) return fs.readFileSync(README_PATH, "utf-8");
  return "# Hack4Good\n\n";
}

// Extract simple children safely from parsed XML
function firstText(obj, key) {
  if (!obj || typeof obj !== "object") return "";
  const v = obj[key];
  if (v === null || v === undefined) return "";
  return String(v).trim();
}

function parseXmlFile(filePath) {
  try {
    const xml = fs.readFileSync(filePath, "utf-8");
    const parser = new XMLParser({
      ignoreAttributes: false,
      allowBooleanAttributes: true
    });
    const root = parser.parse(xml);

    // Navigate to <record_update><x_snc_hack4good_0_hack4good_proposal>...
    const record =
      root?.record_update?.x_snc_hack4good_0_hack4good_proposal ||
      root?.x_snc_hack4good_0_hack4good_proposal;

    if (!record) return null;

    const projectName = firstText(record, "project_name");
    if (!projectName) return null;

    const focusArea = friendlyFocus(firstText(record, "focus_area"));
    const createdRaw = firstText(record, "sys_created_on");
    const createdDt = createdRaw ? new Date(createdRaw.replace(" ", "T") + "Z") : null;

    return {
      project_name: projectName,
      focus_area: focusArea,
      created_dt: createdDt,       // Date | null
      created_raw: createdRaw,     // original string
      path: filePath
    };
  } catch (e) {
    console.error(`WARNING: Failed to parse ${filePath}: ${e.message}`);
    return null;
  }
}

// Get first commit that added the file (like git log --diff-filter=A)
async function firstAddingCommitSha(filePath) {
  // Use `git log` via child_process to avoid extra deps
  // (Actions runners have git)
  const { spawnSync } = await import("child_process");
  const out = spawnSync("git", ["log", "--diff-filter=A", "--format=%H", "--", filePath], {
    encoding: "utf-8"
  });
  if (out.status !== 0) return null;
  const lines = out.stdout.trim().split("\n").filter(Boolean);
  return lines[0] || null;
}

// Resolve PR author for a commit; fallback to commit author
async function resolveAttributionForCommit(commitSha) {
  if (!commitSha || !OWNER || !REPO) return null;

  try {
    // Preferred: list PRs associated with a commit
    const pulls = await octokit.request("GET /repos/{owner}/{repo}/commits/{ref}/pulls", {
      owner: OWNER,
      repo: REPO,
      ref: commitSha,
      mediaType: { format: "json" }
    });

    if (Array.isArray(pulls.data) && pulls.data.length) {
      const pr = pulls.data.sort((a, b) => new Date(a.created_at) - new Date(b.created_at))[0];
      const user = pr.user || {};
      return {
        login: user.login || null,
        avatar_url: user.avatar_url ? `${user.avatar_url}&s=40` : "",
        html_url: user.html_url || ""
      };
    }
  } catch (e) {
    // fall through to commit author
  }

  try {
    const commit = await octokit.repos.getCommit({ owner: OWNER, repo: REPO, ref: commitSha });
    const authorUser = commit.data.author; // linked GitHub user if available
    if (authorUser) {
      return {
        login: authorUser.login || null,
        avatar_url: authorUser.avatar_url ? `${authorUser.avatar_url}&s=40` : "",
        html_url: authorUser.html_url || ""
      };
    }
    const raw = commit.data.commit?.author || {};
    const login = raw.name || raw.email || "Unknown";
    return { login, avatar_url: "", html_url: "" };
  } catch {
    return null;
  }
}

function renderSubmitterCell(attr) {
  const login = attr?.login || "unknown";
  const url = attr?.html_url || "";
  const avatar = attr?.avatar_url || "";
  if (url) {
    const img = avatar
      ? `<img src="${avatar}" width="20" height="20" alt="@${login}"/>`
      : "";
    // Compact inline: avatar then @handle, all inside one link
    return `<a href="${url}">${img} @${login}</a>`;
  }
  return `@${login}`;
}

function replaceBetweenMarkers(readmeText, newBlock) {
  if (!readmeText.includes(MARKER_START) || !readmeText.includes(MARKER_END)) {
    return (
      readmeText.trimEnd() +
      `

${MARKER_START}

_Updated automatically on merge to \`main\`._

${newBlock}
${MARKER_END}
`
    );
  }
  const pattern = new RegExp(
    `(${escapeRegExp(MARKER_START)})([\\s\\S]*?)(${escapeRegExp(MARKER_END)})`,
    "m"
  );
  const replacement = `$1

_Updated automatically on merge to \`main\`._
${newBlock}$3`;
  return readmeText.replace(pattern, replacement);
}

function escapeRegExp(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

async function buildTable(items) {
  if (!items.length) {
    return "\n_No ideas yet. Be the first to submit one!_\n";
  }

  const cache = loadCache();
  const rows = [];

  try {
    for (const it of items) {
      const project = `[${it.project_name}](${it.path})`;
      const focus = it.focus_area;
      const key = it.path;

      let attr = cache[key];
      if (!attr || !attr.login) {
        const sha = await firstAddingCommitSha(it.path);
        attr = await resolveAttributionForCommit(sha);
        cache[key] = attr || { login: "Unknown", avatar_url: "", html_url: "" };
      }

      const submitter = renderSubmitterCell(cache[key]);
      const created = it.created_dt
        ? it.created_dt.toISOString().slice(0, 10)
        : "—";

      rows.push(`| ${project} | ${focus} | ${submitter} | ${created} |`);
    }
  } finally {
    saveCache(cache);
  }

  const header =
    "| Project | Focus area | Submitted by | Created (UTC) |\n|---|---|---|---|\n";
  return "\n" + header + rows.join("\n") + "\n";
}

async function main() {
  // 1) Find XMLs
  const files = await fg(GLOB_PATTERN, { dot: true, onlyFiles: true });
  const items = [];

  for (const f of files) {
    const rec = parseXmlFile(f);
    if (rec) items.push(rec);
  }

  // 2) Sort newest first (entries with no date sink)
  items.sort((a, b) => {
    const av = a.created_dt ? a.created_dt.getTime() : -Infinity;
    const bv = b.created_dt ? b.created_dt.getTime() : -Infinity;
    return bv - av;
  });

  // 3) Build table (Markdown)
  const tableMd = await buildTable(items);

  // 4) Inject into README
  const readme = readReadme();
  const updated = replaceBetweenMarkers(readme, tableMd);

  if (updated !== readme) {
    fs.writeFileSync(README_PATH, updated, "utf-8");
    console.log("README.md updated.");
  } else {
    console.log("README.md unchanged.");
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/update_readme.py
#!/usr/bin/env python3
from pathlib import Path
import xml.etree.ElementTree as ET
from datetime import datetime
import re
import sys
import os
import json
import subprocess
import urllib.request
import urllib.error

MARKER_START = "<!-- ideas:start -->"
MARKER_END = "<!-- ideas:end -->"
README_PATH = Path("README.md")
ATTR_CACHE_PATH = Path(".idea_attribution.json")

# Focus area mapping
FOCUS_MAP = {
    "cure": "🧪 Cause and Cure",
    "disaster": "🆘 Disaster and Community Support",
    "education": "🎓 Education and Youth Services",
    "sustainability": "🌱 Sustainability and Decarbonization",
    "other": "🧩 Other",
}

GLOB_PATTERN = "**/update/x_snc_hack4good_0_hack4good_proposal_*.xml"

def friendly_focus(value: str) -> str:
    if not value:
        return "—"
    return FOCUS_MAP.get(value.strip().lower(), value.replace("_", " ").title())

def parse_xml(file: Path):
    try:
        root = ET.parse(file).getroot()
        rec = root.find(".//x_snc_hack4good_0_hack4good_proposal")
        if rec is None:
            return None
        def text(tag):
            el = rec.find(tag)
            return (el.text or "").strip() if el is not None else ""
        project_name = text("project_name")
        focus_area = friendly_focus(text("focus_area"))
        created_raw = text("sys_created_on")
        created_dt = None
        if created_raw:
            created_dt = datetime.strptime(created_raw, "%Y-%m-%d %H:%M:%S")
        if not project_name:
            return None
        return {
            "project_name": project_name,
            "focus_area": focus_area,
            "created_dt": created_dt,
            "created_raw": created_raw,
            "path": str(file),
        }
    except Exception as e:
        print(f"WARNING: Failed to parse {file}: {e}", file=sys.stderr)
        return None

def git_first_commit_for_path(path: str) -> str | None:
    try:
        sha = subprocess.check_output(
            ["git", "log", "--diff-filter=A", "--format=%H", "--", path],
            text=True
        ).strip().splitlines()
        return sha[0] if sha else None
    except subprocess.CalledProcessError:
        return None

def gh_api(url: str):
    token = os.environ.get("GITHUB_TOKEN") or os.environ.get("GH_TOKEN")
    headers = {
        "User-Agent": "hack4good-readme-bot",
        "Accept": "application/vnd.github+json",
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=20) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        print(f"GitHub API error {e.code} for {url}", file=sys.stderr)
    except Exception as e:
        print(f"GitHub API exception for {url}: {e}", file=sys.stderr)
    return None

def pr_author_for_commit(owner: str, repo: str, commit_sha: str):
    """
    Use: GET /repos/{owner}/{repo}/commits/{commit_sha}/pulls
    Fallback: GET /repos/{owner}/{repo}/commits/{sha} to get commit author.
    """
    if not commit_sha:
        return None
    pulls = gh_api(f"https://api.github.com/repos/{owner}/{repo}/commits/{commit_sha}/pulls")
    if isinstance(pulls, list) and pulls:
        # pick the earliest merged PR or the first
        pulls_sorted = sorted(
            pulls, key=lambda p: (p.get("merged_at") or p.get("created_at") or ""), reverse=False
        )
        pr = pulls_sorted[0]
        user = pr.get("user") or {}
        return {
            "login": user.get("login"),
            "avatar_url": (user.get("avatar_url") or "") + "&s=40",
            "html_url": user.get("html_url"),
        }
    # Fallback to commit author info
    commit = gh_api(f"https://api.github.com/repos/{owner}/{repo}/commits/{commit_sha}")
    if isinstance(commit, dict):
        # Prefer linked GitHub user if present
        if commit.get("author"):
            user = commit["author"]
            return {
                "login": user.get("login"),
                "avatar_url": (user.get("avatar_url") or "") + "&s=40",
                "html_url": user.get("html_url"),
            }
        # Otherwise use the raw commit author (name/email), without avatar
        raw = commit.get("commit", {}).get("author", {}) or {}
        login = raw.get("name") or raw.get("email") or "Unknown"
        return {"login": login, "avatar_url": "", "html_url": ""}
    return None

def load_cache():
    if ATTR_CACHE_PATH.exists():
        try:
            return json.loads(ATTR_CACHE_PATH.read_text(encoding="utf-8"))
        except Exception:
            pass
    return {}

def save_cache(cache):
    ATTR_CACHE_PATH.write_text(json.dumps(cache, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")

def resolve_attribution(xml_path: str, cache: dict, owner: str, repo: str):
    if xml_path in cache and cache[xml_path].get("login"):
        return cache[xml_path]
    first_sha = git_first_commit_for_path(xml_path)
    info = pr_author_for_commit(owner, repo, first_sha) if first_sha else None
    cache[xml_path] = info or {"login": "Unknown", "avatar_url": "", "html_url": ""}
    return cache[xml_path]

def render_submitter(cell):
    login = cell.get("login") or "unknown"
    url = cell.get("html_url") or ""
    avatar = cell.get("avatar_url") or ""
    if url:
        img = f'<img src="{avatar}" width="20" height="20" style="border-radius:50%; vertical-align:middle;" alt="@{login}"/>'
        return f'<a href="{url}">{img} @{login}</a>'
    return f"@{login}"

def build_table(items, owner: str, repo: str):
    if not items:
        return "\n_No ideas yet. Be the first to submit one!_\n"

    # Column order: Project | Focus area | Avatar | Created
    header = "| Project | Focus area | Submitted by | Created (UTC) |\n|---|---|---|---|\n"
    rows = []
    cache = load_cache()
    try:
        for it in items:
            project = f"[{it['project_name']}]({it['path']})"
            focus = it['focus_area']
            submitter = resolve_attribution(it["path"], cache, owner, repo)
            submitter_cell = render_submitter(submitter)
            created = it["created_dt"].strftime("%Y-%m-%d") if it["created_dt"] else "—"
            rows.append(f"| {project} | {focus} | {submitter_cell} | {created} |")
    finally:
        save_cache(cache)

    return "\n" + header + "\n".join(rows) + "\n"

def replace_between_markers(readme_text: str, new_block: str) -> str:
    if MARKER_START not in readme_text or MARKER_END not in readme_text:
        return (
            readme_text.rstrip()
            + f"\n\n{MARKER_START}\n\n_Updated automatically on merge to `main`._\n\n"
            + new_block
            + f"\n{MARKER_END}\n"
        )
    pattern = re.compile(
        rf"({re.escape(MARKER_START)})(.*)({re.escape(MARKER_END)})",
        flags=re.DOTALL,
    )
    replacement = rf"\1\n\n_Updated automatically on merge to `main`._\n{new_block}\3"
    return pattern.sub(replacement, readme_text)

def main():
    repo_full = os.environ.get("GITHUB_REPOSITORY", "")
    owner, repo = (repo_full.split("/", 1) + [""])[:2]

    files = sorted(Path(".").glob(GLOB_PATTERN))
    items = []
    for f in files:
        rec = parse_xml(f)
        if rec:
            items.append(rec)

    items.sort(key=lambda x: x["created_dt"] or datetime.min, reverse=True)
    table_md = build_table(items, owner, repo)

    readme = README_PATH.read_text(encoding="utf-8") if README_PATH.exists() else "# Hack4Good\n\n"
    updated = replace_between_markers(readme, table_md)

    if updated != readme:
        README_PATH.write_text(updated, encoding="utf-8")
        print("README.md updated.")
    else:
        print("README.md unchanged.")

if __name__ == "__main__":
    main()
